Update lib to work with prices/django-prices beta 3

rafalp opened the pull request:
This PR is followup for changes that occured in prices 1.0.0b3 and django-prices. In addition this PR adds extra tests for better coverage as well as synces our requirements and tests config with prices and django prices.

**TODO**

* [x] Make tests pass
* [x] Sync requirements with other libs
* [x] Test `discount_amount_in_currency`
* [x] Rename 'value' and 'price' to 'base' whenever value's usage is unknown or too generic
* [x] Remove `from __future__` imports
* [x] Try refactoring `convert_price`
* [x] Add `mypy` type annotations wherever possible
* [x] Remove `prices_multicurrency_i18n` filters
* [x] Remove `discount_amount_in_currency` filter
* [x] Cut down number of times that  `exchange_currency` fetches conversion rate for types other than `Money`

patrys commented on django_prices_openexchangerates/__init__.py:
```diff
@@ -16,7 +16,7 @@
 class CurrencyConversion(object):
```
In Python 3 `(object)` is no more.

patrys commented on django_prices_openexchangerates/__init__.py:
```diff
@@ -29,14 +29,14 @@ def __repr__(self):
+            return Money(
+                price_obj.amount * self.rate, currency=self.to_currency)
+        if isinstance(price_obj, TaxedMoney):
+            return TaxedMoney(Money(price_obj.net.amount * self.rate,
```
Please use normal wrap after the opening parenthesis.

patrys commented on django_prices_openexchangerates/templatetags/prices_multicurrency_i18n.py:
```diff
@@ -8,18 +8,18 @@
-    discount_amount = exchange_currency(discount.amount, to_currency=currency)
-    discount.amount = discount_amount
-    return (amount | discount) - amount
+def discount_amount_in_currency(discount, price, currency):
```
I think `base` is better than `price`. Not sure whether we should keep this order of arguments.

rafalp replied:
The arguments here are mostly like that for consistency with `django-prices`'s `discount_amount_for(discount, price)` template tag.

patrys commented on django_prices_openexchangerates/__init__.py:
```diff
@@ -53,7 +53,7 @@ def get_conversion_rate(currency):
 
 def convert_price(price, to_currency, get_rate=get_conversion_rate):
```
What is a `price` here (both in function name and in signature)?

rafalp commented:
@patrys I've tried moving `convert_price`'s logic  around. I'm still unsure if we want to keep the `CurrencyConversion` around, or perhaps extract its `apply` logic to function/move it to `exchange_currency`.

patrys commented:
I would drop it for the same reason we've dropped discount and tax classes. It's just as easy to use `functools.partial` to achieve the same effect.

rafalp commented:
Removed the `CurrencyConversion` and moved it's apply logic to `exchange_currency`. Seeing how we are py3k only now, I've also removed `__future__` form codebase. If we are on the mark here, I'll move on to adding type annotations for `mypy`.

patrys commented on django_prices_openexchangerates/__init__.py:
```diff
@@ -1,45 +1,19 @@
+from prices import Money, MoneyRange, TaxedMoney, TaxedMoneyRange
 
-BASE_CURRENCY = getattr(settings, 'OPENEXCHANGERATES_BASE_CURRENCY', 'USD')
+Conversion = Union[float, int, Decimal]
```
Are you sure we want to support `float`?

patrys commented on django_prices_openexchangerates/__init__.py:
```diff
@@ -72,22 +45,38 @@ def convert_price(price, to_currency, get_rate=get_conversion_rate):
+        return Money(base.amount * conversion_rate, currency=to_currency)
+    if isinstance(base, MoneyRange):
+        return MoneyRange(exchange_currency(base.start, to_currency,
+                                            get_rate=get_rate),
```
Please don't do line continuation like that. Fold after each opening parenthesis.

patrys commented on django_prices_openexchangerates/__init__.py:
```diff
@@ -72,22 +45,38 @@ def convert_price(price, to_currency, get_rate=get_conversion_rate):
+        return TaxedMoneyRange(exchange_currency(base.start, to_currency,
+                                                 get_rate=get_rate),
+                               exchange_currency(base.stop, to_currency,
+                                                 get_rate=get_rate))
```
What if none of the conditionals match?

patrys commented on django_prices_openexchangerates/migrations/0001_initial.py:
```diff
@@ -1,17 +1,12 @@
 # -*- coding: utf-8 -*-
 # Generated by Django 1.9.2 on 2016-03-17 09:36
-from __future__ import unicode_literals
```
Are you sure that editing migrations is a good idea?

rafalp replied:
I've tested this change locally with Django 2.0.1 on py35k and py36k and can confirm that `makemigrations` reports `No changes detected` after the removal of `__future__` import, something I'm consider farm more acceptable than editing `operations` content.

patrys commented on django_prices_openexchangerates/templatetags/prices_multicurrency_i18n.py:
```diff
@@ -1,27 +1,32 @@
-    discount.amount = discount_amount
-    return (amount | discount) - amount
+def discount_amount_in_currency(
+    base: T, discount: Callable[[T], T], currency: str) -> T:
```
Bad indentation.

patrys commented on setup.py:
```diff
@@ -33,7 +32,8 @@
     classifiers=CLASSIFIERS,
-    install_requires=['Django>=1.4', 'django-prices>=1.0.0-beta2', 'prices>=1.0.0-beta2'],
+    install_requires=[
+        'Django<2.1.0', 'django-prices>=1.0.0-beta3', 'prices>=1.0.0-beta3'],
```
Why `Django<2.1.0` instead of `Django>=1.11`?

patrys commented on django_prices_openexchangerates/__init__.py:
```diff
@@ -1,45 +1,19 @@
-
-
-def get_conversion_rate(currency):
+def get_rate_from_db(currency: str) -> Conversion:
```
Why does it return a union here? Does the returned type differ between calls?

patrys commented on django_prices_openexchangerates/__init__.py:
```diff
@@ -51,43 +25,57 @@ def get_conversion_rate(currency):
-def convert_price(price, to_currency, get_rate=get_conversion_rate):
+def get_conversion_rate(
+        from_currency: str, to_currency: str,
+        get_rate: Callable[[str], Conversion]) -> Conversion:
```
Same, why return a union type?

patrys commented on django_prices_openexchangerates/__init__.py:
```diff
@@ -51,43 +25,57 @@ def get_conversion_rate(currency):
+            exchange_currency(base.stop, to_currency, get_rate=get_rate))
+    if isinstance(base, TaxedMoney):
+        return TaxedMoney(
+            Money(base.net.amount * conversion_rate, currency=to_currency),
```
A recursive call would make it more readable.

patrys commented on django_prices_openexchangerates/templatetags/prices_multicurrency.py:
```diff
@@ -1,19 +1,23 @@
-    return converted_amount
+def in_currency(base: T, currency: str) -> T:
+    converted_base = exchange_currency(base, currency)
+    converted_base = converted_base.quantize('.01')
```
This quantize call is wrong. We can't assume two decimal places.

patrys commented on django_prices_openexchangerates/templatetags/prices_multicurrency_i18n.py:
```diff
@@ -1,27 +1,32 @@
-    return converted_amount
+def in_currency(base: T, currency: str) -> T:
+    converted_base = exchange_currency(base, currency)
+    converted_base = converted_base.quantize('.01')
```
Also wrong.

patrys commented on django_prices_openexchangerates/templatetags/prices_multicurrency_i18n.py:
```diff
@@ -1,27 +1,32 @@
-def amount(obj, format='text'):
-    return prices_i18n.amount(obj, format=format)
+def amount(base: Money, format='text') -> str:
+    return prices_i18n.amount(base, format=format)
```
Why is this templatetag useful?

rafalp commented:
Another pass on our codebase:

- Removed `discount_amount_in_currency` because it didn't fit into mental image of what lib should do (currency conversion, not math)
- Removed `i18n` version of templatetags because its codebase was identical to non-i18n module
- Recurrent calls inside `exchange_currency` reuse `conversion_rate`, greatly shaving number of hits to outside resources for complex money types like `TaxedMoneyRange`.
- Cleanup the tests suite from tests that claimed to test different things but in fact tested same inputs and results.
- Address some style issues raised by `pycodestyle`, such as double blank lines in otherwhise empty files
